File: api/app/services/tenant_service.py
"""
Tenant service for tenant management.
"""
from typing import List, Optional
from sqlalchemy.orm import Session

from app.services.base import BaseService
from app.repositories.tenant_repository import TenantRepository
from app.models.tenant import Tenant
from app.schemas.tenant import TenantCreate, TenantUpdate


class TenantService(BaseService):
    """Service for tenant management."""

    def __init__(self, db: Session):
        super().__init__(db)
        self.tenant_repo = TenantRepository(db)

    def list_tenants(self, org_id: int) -> List[Tenant]:
        """List all tenants in an organization."""
        return self.tenant_repo.find_by_organization(org_id)

    def get_tenant(self, tenant_id: int, org_id: int) -> Optional[Tenant]:
        """Get a tenant by ID within an organization."""
        tenant = self.tenant_repo.get(tenant_id)
        if tenant and tenant.organization_id == org_id:
            return tenant
        return None

    def create_tenant(self, org_id: int, data: TenantCreate) -> Tenant:
        """Create a new tenant."""
        tenant = Tenant(
            organization_id=org_id,
            name=data.name,
            phone=data.phone,
            email=data.email,
            id_card=data.id_card,
            emergency_contact=data.emergency_contact,
            emergency_phone=data.emergency_phone,
            notes=data.notes,
        )
        return self.tenant_repo.create(tenant)

    def update_tenant(
        self, tenant_id: int, org_id: int, data: TenantUpdate
    ) -> Optional[Tenant]:
        """Update a tenant."""
        tenant = self.get_tenant(tenant_id, org_id)
        if not tenant:
            return None
        update_data = data.model_dump(exclude_unset=True)
        return self.tenant_repo.update(tenant_id, **update_data)

    def delete_tenant(self, tenant_id: int, org_id: int) -> bool:
        """Delete a tenant."""
        tenant = self.get_tenant(tenant_id, org_id)
        if not tenant:
            return False
        return self.tenant_repo.delete(tenant_id)

File: api/src/services/tenant.service.ts
import type { Tenant, Prisma } from '@prisma/client';
import { ulid } from 'ulid';
import { createTenantRepository, type TenantRepository } from '../repositories/tenant.repo.js';
import { createAppError } from '../utils/appError.js';
import { NotFoundMessages } from '../messages.js';
import { prisma } from '../lib/prisma.js';
import { compareNaturalText } from '../utils/intuitiveSort.js';

/**
 * 创建租客输入
 */
export interface CreateTenantInput {
  name: string;
  phone?: string;
  sms_opt_out?: boolean;
  sms_opt_out_reason?: string;
  id_card?: string;
  emergency_contact?: string;
  emergency_phone?: string;
  notes?: string;
}

/**
 * 更新租客输入
 */
export interface UpdateTenantInput {
  name?: string;
  phone?: string;
  sms_opt_out?: boolean;
  sms_opt_out_reason?: string;
  id_card?: string;
  emergency_contact?: string;
  emergency_phone?: string;
  notes?: string;
}

/**
 * 构建租客创建数据
 */
function buildCreateData(orgId: string, data: CreateTenantInput): Prisma.TenantCreateInput {
  const smsOptOut = data.sms_opt_out ?? false;
  return {
    id: ulid().toLowerCase(),
    organization: { connect: { id: orgId } },
    name: data.name,
    phone: data.phone,
    sms_opt_out: smsOptOut,
    sms_opt_out_at: smsOptOut ? new Date() : null,
    sms_opt_out_reason: smsOptOut ? data.sms_opt_out_reason : null,
    id_card: data.id_card,
    emergency_contact: data.emergency_contact,
    emergency_phone: data.emergency_phone,
    notes: data.notes,
  };
}

/**
 * 构建租客更新数据
 */
function buildUpdateData(existing: Tenant, data: UpdateTenantInput): Prisma.TenantUpdateInput {
  const smsOptOut = data.sms_opt_out ?? existing.sms_opt_out;
  const smsOptOutChanged = data.sms_opt_out != null && data.sms_opt_out !== existing.sms_opt_out;
  return {
    name: data.name ?? existing.name,
    phone: data.phone ?? existing.phone,
    sms_opt_out: smsOptOut,
    sms_opt_out_at: smsOptOut
      ? smsOptOutChanged
        ? new Date()
        : existing.sms_opt_out_at
      : null,
    sms_opt_out_reason: smsOptOut
      ? data.sms_opt_out_reason ?? existing.sms_opt_out_reason
      : null,
    id_card: data.id_card ?? existing.id_card,
    emergency_contact: data.emergency_contact ?? existing.emergency_contact,
    emergency_phone: data.emergency_phone ?? existing.emergency_phone,
    notes: data.notes ?? existing.notes,
  };
}

/**
 * Tenant Service 接口
 */
export interface TenantService {
  list(orgId: string, search?: string): Promise<Tenant[]>;
  getById(orgId: string, id: string): Promise<Tenant>;
  getByIds(ids: string[]): Promise<Tenant[]>;
  create(orgId: string, data: CreateTenantInput): Promise<Tenant>;
  update(orgId: string, id: string, data: UpdateTenantInput): Promise<Tenant>;
  delete(orgId: string, id: string): Promise<void>;
}

/**
 * 创建 Tenant Service 实例
 * @param getRepo - Repository 工厂函数，支持依赖注入
 */
export function createTenantService(
  getRepo: () => TenantRepository = () => createTenantRepository(prisma)
): TenantService {
  return {
    list: async (orgId: string, search?: string) => {
      const tenants = await getRepo().findByOrgId(orgId, search);
      return [...tenants].sort(
        (left, right) =>
          compareNaturalText(left.name, right.name) ||
          compareNaturalText(left.phone, right.phone)
      );
    },

    getById: async (orgId: string, id: string) => {
      const tenant = await getRepo().findByIdAndOrg(id, orgId);
      if (!tenant) {
        throw createAppError(404, NotFoundMessages.TENANT);
      }
      return tenant;
    },

    getByIds: async (ids: string[]) => {
      return getRepo().findByIds(ids);
    },

    create: async (orgId: string, data: CreateTenantInput) => {
      return getRepo().create(buildCreateData(orgId, data));
    },

    update: async (orgId: string, id: string, data: UpdateTenantInput) => {
      const existing = await getRepo().findByIdAndOrg(id, orgId);
      if (!existing) {
        throw createAppError(404, NotFoundMessages.TENANT);
      }
      return getRepo().update(id, buildUpdateData(existing, data));
    },

    delete: async (orgId: string, id: string) => {
      const existing = await getRepo().findByIdAndOrg(id, orgId);
      if (!existing) {
        throw createAppError(404, NotFoundMessages.TENANT);
      }
      await getRepo().delete(id);
    },
  };
}

/**
 * 默认 Tenant Service 实例
 */
export const defaultTenantService = createTenantService();
